feature: add registry authorizer retry

TerryHowe opened the pull request:
**What this PR does / why we need it**:

I have functionally tested and docker.io uses bearer auth and ghcr.io uses basic auth. This should fix the Quay issue and the private registry issue.

Closes: https://github.com/helm/helm/issues/31256
Closes: https://github.com/helm/helm/issues/31567

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

TerryHowe committed:
feature: add registry authorizer retry

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

benoittgt reviewed (commented):
Thanks! And sorry again for the extra work.   I was wondering if we should consider thread safety for flags like `AttemptBearerAuthentication`?  Also, do we need Go doc ?

TerryHowe committed:
simplify tests

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
chore: handle mutliple threads

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
chore: handle specific errors

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
fix: address authorizer review comments

- Normalize ghcr.io hostname using net.SplitHostPort + strings.EqualFold
- Save and restore ForceAttemptOAuth2 via defer so it is always false
  after Do() returns, fixing regression tests
- Explicitly switch to basic auth (ForceAttemptOAuth2=false) before
  retrying after a 401/403 error so the second attempt uses a different
  auth mode
- Fix require.NoError called from goroutine in concurrent test; use
  t.Errorf+return instead (testify require is not goroutine-safe)
- Replace TestAuthorizer_Do_StatusCodeErrorChecking with tests that
  exercise Authorizer.Do directly: RetriesOn401 and NoRetryOn404
- Restore TestLogin_ResetsForceAttemptOAuth2_OnSuccess and
  TestLogin_ResetsForceAttemptOAuth2_OnFailure regression tests that
  were removed; these now pass with the defer-based restore

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on pkg/registry/authorizer_test.go:
```diff
@@ -0,0 +1,345 @@
+			defer wg.Done()
+			for j := 0; j < numRequests; j++ {
+				req, err := http.NewRequest(http.MethodGet, server.URL, nil)
+				require.NoError(t, err)
```
Fixed — changed to `t.Errorf("failed to create request: %v", err); return` which is safe to call from goroutines.

TerryHowe commented on pkg/registry/authorizer_test.go:
```diff
@@ -0,0 +1,345 @@
+	tests := []struct {
+		name        string
+		errorMsg    string
+		shouldRetry bool
```
Replaced TestAuthorizer_Do_StatusCodeErrorChecking with two tests that exercise Authorizer.Do directly via a counting RoundTripper: TestAuthorizer_Do_RetriesOn401 (verifies a second attempt is made after a 401 error) and TestAuthorizer_Do_NoRetryOn404 (verifies no retry on non-auth errors).

TerryHowe commented on pkg/registry/authorizer.go:
```diff
@@ -0,0 +1,107 @@
+			a.setForceAttemptOAuth2(true)
+			if originalReq.Host == "ghcr.io" {
+				a.setForceAttemptOAuth2(false)
+				a.setAttemptBearerAuthentication(false)
```
Fixed — added an isGHCR() helper that uses net.SplitHostPort + strings.EqualFold to normalize the host before comparing, so both 'ghcr.io' and 'ghcr.io:443' (or any case variant) are handled correctly.

TerryHowe commented on pkg/registry/authorizer.go:
```diff
@@ -0,0 +1,107 @@
+			if originalReq.Host == "ghcr.io" {
+				a.setForceAttemptOAuth2(false)
+				a.setAttemptBearerAuthentication(false)
+			}
```
Fixed — for the non-ghcr.io branch, we now save the previous value of ForceAttemptOAuth2 before setting it to true, and register a defer to restore it when Do() returns. This ensures ForceAttemptOAuth2 is always back to its prior value (false by default) after any exit path from Do(), which is what the restored regression tests verify.

TerryHowe commented on pkg/registry/authorizer.go:
```diff
@@ -0,0 +1,107 @@
+			if !strings.Contains(err.Error(), "response status code 401") &&
+				!strings.Contains(err.Error(), "response status code 403") {
+				return nil, err
+			}
```
Fixed — added an explicit setForceAttemptOAuth2(false) before the fallback retry so the second attempt uses basic auth rather than repeating the same OAuth2/bearer attempt.

TerryHowe commented on pkg/registry/client_test.go:
```diff
@@ -56,71 +52,6 @@ func TestTagManifestTransformsReferences(t *testing.T) {
-	}
-}
-
 // TestWarnIfHostHasPath verifies that warnIfHostHasPath correctly detects path components.
 func TestWarnIfHostHasPath(t *testing.T) {
 	t.Parallel()
```
Restored both TestLogin_ResetsForceAttemptOAuth2_OnSuccess and TestLogin_ResetsForceAttemptOAuth2_OnFailure. Both now pass with the defer-based ForceAttemptOAuth2 restoration.

TerryHowe commented on pkg/registry/authorizer.go:
```diff
@@ -0,0 +1,107 @@
+			resp, err := a.Client.Do(originalReq)
+			if err == nil {
+				a.setAttemptBearerAuthentication(false)
+				return resp, nil
```
Per-host tracking is a larger refactor that would change the Authorizer's interface significantly. Left as a follow-up; the current single-flag approach is consistent with how the client is typically used (one registry per client instance for push/pull operations).

TerryHowe commented on pkg/registry/authorizer.go:
```diff
@@ -0,0 +1,107 @@
+			}
+			if !strings.Contains(err.Error(), "response status code 401") &&
+				!strings.Contains(err.Error(), "response status code 403") {
+				return nil, err
```
Acknowledged. The string 'response status code 401' matches the format used by oras-go's ErrorResponse type. Switching to typed errors would require depending on oras-go internals; left as a follow-up if oras-go exports a typed error for this in a future version.

TerryHowe commented on pkg/registry/client.go:
```diff
@@ -250,18 +237,12 @@ func (c *Client) Login(host string, options ...LoginOption) error {
-		if err := reg.Ping(ctx); err != nil {
-			return fmt.Errorf("authenticating to %q: %w", host, err)
-		}
+		return fmt.Errorf("authenticating to %q: %w", host, err)
```
The retry/fallback is now handled in Authorizer.Do rather than in Login. When Login calls reg.Ping, the Authorizer.Do method will attempt bearer first and fall back to basic if a 401/403 error is returned, which covers both bearer-only and basic-only registries without duplicating the logic in Login.

TerryHowe committed:
style: fix lint issues in authorizer_test.go

Use range-over-int for loops and rename unused parameters to _.

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
fix(registry): use URL.Host for ghcr.io check and retry with OAuth2 on 401/403

- reqHost() prefers req.URL.Hostname() over req.Host so the ghcr.io
  special case fires correctly for ORAS-constructed requests where the
  host lives in req.URL, not req.Host
- Add dedicated TestAuthorizer_Do_GHCRSkipsBearerProbe test that sets
  the host via URL (as ORAS does) to verify the ghcr.io path
- Redesign Do() retry logic: first attempt uses standard auth; only
  after a 401/403 do we retry with ForceAttemptOAuth2=true to support
  registries whose token endpoints require OAuth2-style requests
  (previously setting it before the first attempt caused 400 errors on
  standard registries)
- Disable attemptBearerAuthentication after successful fallback retry
  so basic-auth-only registries pay the probe cost only once
- Add GoDoc comments on Authorizer, NewAuthorizer, and EnableCache

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on pkg/registry/client.go:
```diff
@@ -292,10 +273,10 @@ func LoginOptPlainText(isPlainText bool) LoginOption {
 
-func ensureTLSConfig(client *auth.Client, setConfig *tls.Config) (*tls.Config, error) {
+func ensureTLSConfig(client *Authorizer, setConfig *tls.Config) (*tls.Config, error) {
 	var transport *http.Transport
 
-	switch t := client.Client.Transport.(type) {
+	switch t := client.Client.Client.Transport.(type) {
 	case *http.Transport:
```
Done.

TerryHowe commented on pkg/registry/authorizer.go:
```diff
@@ -0,0 +1,122 @@
+			a.setForceAttemptOAuth2(false)
+		}
+	}
+	return a.Client.Do(originalReq)
```
Done.

TerryHowe commented on pkg/registry/authorizer.go:
```diff
@@ -0,0 +1,122 @@
+			if isGHCR(originalReq.Host) {
+				a.setForceAttemptOAuth2(false)
+				a.setAttemptBearerAuthentication(false)
+			} else {
```
Done.

TerryHowe commented on pkg/registry/authorizer.go:
```diff
@@ -0,0 +1,122 @@
+	"context"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+
+	"helm.sh/helm/v4/internal/version"
+
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+)
+
+type Authorizer struct {
+	auth.Client
+	lock                        sync.RWMutex
+	attemptBearerAuthentication bool
+}
+
+// isGHCR reports whether host refers to ghcr.io, normalizing port and case.
+func isGHCR(host string) bool {
+	h, _, err := net.SplitHostPort(host)
+	if err != nil {
+		h = host
+	}
+	return strings.EqualFold(h, "ghcr.io")
+}
+
+func NewAuthorizer(httpClient *http.Client, credentialsStore credentials.Store, username, password string) *Authorizer {
+	authorizer := Authorizer{
```
Done.

TerryHowe commented:
nvm
